### DISCUSSION TITLE
how to use ready-notification funciton?

### DISCUSSION THREAD

**Discussion by lfxx** [davmac314/dinit#252] [General]
Hello.I try to use ready-notification funtion.But it is not working.Can you post a demo config here?it would be nicer if you can eleborate this function more clear.at this time,i think it has the following usage:
For example, there are two services, A and B, with B depending on A. I wish to set a variable such that when A changes it, it implies that A has fully started, and at that point, B is started.
am i right?

**Comment by lfxx** [comment]
Urgent on this issue.Anyone come to help me out,thanks!

**Comment by mobin-2008** [comment]
Hello. What did you expect from `ready-notification`? It designed to help daemons to notify supervisor about they're ready for actual functionality.
From [`dinit-service(5)`](https://davmac.org/projects/dinit/man-pages-html/dinit-service.5.html) man-page:
> [ready-notification = {pipefd:fd-number | pipevar:env-var-name}](https://davmac.org/projects/dinit/man-pages-html/dinit-service.5.html#ready)
Specifies the mechanism, if any, by which a process service will notify that it is ready (successfully started). If not specified, a process service is considered started as soon as it has begun execution. The two options are:

> pipefd:fd-number — the service will write a message to the specified file descriptor, which dinit sets up as the write end of a pipe before execution. This mechanism is compatible with the S6 supervision suite.

> pipevar:env-var-name — the service will write a message to file descriptor identified using the contents of the specified environment variable, which will be set by dinit before execution to a file descriptor (chosen arbitrarily) attached to the write end of a pipe.

**Comment by davmac314** [comment]
Example setting:

    ready-notification = fd:3

This would set allow the service process to signal readiness by writing to file descriptor number 3.

> For example, there are two services, A and B, with B depending on A. I wish to set a variable such that when A changes it, it implies that A has fully started, and at that point, B is started.

This is roughly what readiness notification is for, yes (if the variable is internal to dinit). When the process for service A indicates readiness, then service A is considered started and then the service B process can be started.

**Reply by lfxx** [reply]
for example,now i has 
A setting:
ready-notification = pipevar:worker_status
B setting:
depends-ms=A
A has a python demo:
```
import time
import os
time.sleep(5)
os.environ.setdefault("worker_status","ok")
```
use `dinit -d dinit_config/ B` can not start B service.How should i do?just want to know how to change the var **worker_status**
i will make my question more clear.
My user case is start A is time consuming.i want to do a funtion test after A start,if pass,then change the var **worker_status** to indicate A is ready.Then we can start B to receive request.Has dinit supported this now?
@mobin-2008 @davmac314

**Reply by SolitudeSF** [reply]
`pipevar` writes file descriptor number to the variable you provided. you dont need to write to the variable, you need to read it, and write a byte to the file descriptor it specifies.

**Reply by lfxx** [reply]
sorry.but i still don't understand.Could you please post a demo code here,thank you very much!i don't how to obtain  the specific file descriptor in python.is this function just for c++ or c?

**Comment by lfxx** [comment]
finally,i use `fd=os.fdopen(3, 'w')` open the file descriptor successfully.

**Reply by davmac314** [reply]
If `worker_status` contains 3, then that is correct approach, yes.

**Reply by davmac314** [reply]
(or if you are using the example setting I provided above.)
